### PR TITLE
Constructable exercise equipment

### PR DESCRIPTION
**Pull request by Gristlebee**
## About The Pull Request

- Punching bags can now be constructed from 5 cloth and deconstructed with a wirecutter.
- Bench presses and chest presses can be constructed from 5 metal and deconstructed with a wrench.
## Why It's Good For The Game

You no longer need to smash apart the exercise equipment like a caveman, and offers players some more freedom in how they want to decorate or arrange their ships.

## Changelog

:cl:
add: Punching bags can be built with 5 cloth.
add: Bench press and chest press can be built with 5 metal.
add: Exercise equipment can be deconstructed.
/:cl:

### CODE REVIEW

**Comment by Apogee-dev** [comment]
I’d actually prefer if exercise equipment could be unanchored and dragged in addition to deconstructed. Maybe allow for unanchoring weight benches with a wrench and deconstructing with a screwdriver?

**Comment by Gristlebee** [comment]
> I’d actually prefer if exercise equipment could be unanchored and dragged in addition to deconstructed. Maybe allow for unanchoring weight benches with a wrench and deconstructing with a screwdriver?

It's been added, and the punching bag can also be unwrenched and wirecuttered

File: code/modules/mining/lavaland/ruins/gym.dm [hunk 1]
```diff
@@ -7,6 +7,51 @@
...
+		if(buildstacktype)
+			new buildstacktype(loc,buildstackamount)
+	..()
+
+/obj/structure/punching_bag/wrench_act(mob/living/user, obj/item/W)
+	if(..())
+		return TRUE
+	add_fingerprint(user)
+	var/action = anchored ? "unbolts [src] from" : "bolts [src] to"
+	var/uraction = anchored ? "unsbolt [src] from" : "bolt [src] to"
```
Comment (west3436) [inline code/modules/mining/lavaland/ruins/gym.dm]: typo "unsbolt"

File: code/modules/mining/lavaland/ruins/gym.dm [hunk 2]
```diff
@@ -7,6 +7,51 @@
...
+	user.visible_message("<span class='warning'>[user] [action] the floor.</span>", "<span class='notice'>You start to [uraction] the floor...</span>", "<span class='hear'>You hear rustling noises.</span>")
+	if(W.use_tool(src, user, 50, volume=100, extra_checks = CALLBACK(src, PROC_REF(check_anchored_state), anchored)))
+		set_anchored(!anchored)
+		to_chat(user, "<span class='notice'>You [anchored ? "unbolt" : "bolt"] [src] from the floor.</span>")
+		return TRUE
+	else
+		return TRUE
```
Comment (west3436) [inline code/modules/mining/lavaland/ruins/gym.dm]: Ensure you want to return TRUE here; if not, remove the "else" statement entirely

File: code/modules/mining/lavaland/ruins/gym.dm [hunk 3]
```diff
@@ -7,6 +7,51 @@
...
+/obj/structure/punching_bag/wirecutter_act(mob/living/user, obj/item/W)
+	. = ..()
+	if(!anchored)
+		user.visible_message("<span class='warning'>[user] cuts apart [src].</span>", "<span class='notice'>You start to cut apart [src].</span>", "<span class='hear'>You hear cutting.</span>")
+		if(W.use_tool(src, user, 50, volume=100))
+			if(anchored)
+				return TRUE
+			to_chat(user, "<span class='notice'>You cut apart [src].</span>")
+			deconstruct(TRUE)
+		return TRUE
```
Comment (west3436) [inline code/modules/mining/lavaland/ruins/gym.dm:40]: Verify that TRUE should be returned here even if the action is not completed.

File: code/modules/mining/lavaland/ruins/gym.dm [hunk 4]
```diff
@@ -33,6 +80,49 @@
...
+		if(buildstacktype)
+			new buildstacktype(loc,buildstackamount)
+	..()
+
+/obj/structure/weightmachine/wrench_act(mob/living/user, obj/item/W)
+	if(..())
+		return TRUE
+	add_fingerprint(user)
+	var/action = anchored ? "unbolts [src] from" : "bolts [src] to"
+	var/uraction = anchored ? "unsbolt [src] from" : "bolt [src] to"
```
Comment (west3436) [inline code/modules/mining/lavaland/ruins/gym.dm]: Same typo as above

File: code/modules/mining/lavaland/ruins/gym.dm [hunk 5]
```diff
@@ -33,6 +80,49 @@
...
+	user.visible_message("<span class='warning'>[user] [action] the floor.</span>", "<span class='notice'>You start to [uraction] the floor...</span>", "<span class='hear'>You hear rustling noises.</span>")
+	if(W.use_tool(src, user, 50, volume=100, extra_checks = CALLBACK(src, PROC_REF(check_anchored_state), anchored)))
+		set_anchored(!anchored)
+		to_chat(user, "<span class='notice'>You [anchored ? "unbolt" : "bolt"] [src] from the floor.</span>")
+		return TRUE
+	else
+		return TRUE
```
Comment (west3436) [inline code/modules/mining/lavaland/ruins/gym.dm]: Same note regarding TRUE returns.

File: code/modules/mining/lavaland/ruins/gym.dm (hunk 1 above)
Comment (Gristlebee) [reply code/modules/mining/lavaland/ruins/gym.dm]: Typo fixed.

**Comment by Gristlebee** [comment]
Typos have been fixed, the extraneous TRUE returns are removed. Also fixed a typo I noticed. Bolt and unbolt messages were backwards.

File: code/modules/mining/lavaland/ruins/gym.dm (hunk 2 above)
Comment (Gristlebee) [reply code/modules/mining/lavaland/ruins/gym.dm]: Not having the TRUEs present causes the mob to attack the bench/bags after the tool actions.

**Comment by Gristlebee** [comment]
Seems like the TRUEs are necessary, but now Rust is failing. Agh.

**Comment by Gristlebee** [comment]
Typos are fixed, the the TRUEs are necessary to prevent the tools from hitting the benches or the bag after the action. Everything should hopefully be good now.

File: code/modules/mining/lavaland/ruins/gym.dm [hunk 6]
```diff
@@ -7,6 +7,51 @@
...
+	add_fingerprint(user)
+	var/action = anchored ? "unbolts [src] from" : "bolts [src] to"
+	var/uraction = anchored ? "unbolt [src] from" : "bolt [src] to"
+	user.visible_message("<span class='warning'>[user] [action] the floor.</span>", "<span class='notice'>You start to [uraction] the floor...</span>", "<span class='hear'>You hear rustling noises.</span>")
+	if(W.use_tool(src, user, 50, volume=100, extra_checks = CALLBACK(src, PROC_REF(check_anchored_state), anchored)))
+		set_anchored(!anchored)
+		to_chat(user, "<span class='notice'>You [anchored ? "bolt" : "unbolt"] [src] from the floor.</span>")
+		return TRUE
+	else
+		return TRUE
```
Comment (MarkSuckerberg) [inline code/modules/mining/lavaland/ruins/gym.dm]:
```suggestion
	return TRUE
```

File: code/modules/mining/lavaland/ruins/gym.dm [hunk 7]
```diff
@@ -7,6 +7,51 @@
...
 	var/list/hit_sounds = list('sound/weapons/genhit1.ogg', 'sound/weapons/genhit2.ogg', 'sound/weapons/genhit3.ogg',\
 	'sound/weapons/punch1.ogg', 'sound/weapons/punch2.ogg', 'sound/weapons/punch3.ogg', 'sound/weapons/punch4.ogg')
+	var/buildstacktype = /obj/item/stack/sheet/cotton/cloth
+	var/buildstackamount = 5
+
+/obj/structure/punching_bag/deconstruct(disassembled = TRUE)
+	if(!(flags_1 & NODECONSTRUCT_1))
+		if(buildstacktype)
+			new buildstacktype(loc,buildstackamount)
+	..()
```
Comment (MarkSuckerberg) [inline code/modules/mining/lavaland/ruins/gym.dm]:
```suggestion
	return ..()
```

File: code/modules/mining/lavaland/ruins/gym.dm [hunk 8]
```diff
@@ -33,6 +80,49 @@
...
+	add_fingerprint(user)
+	var/action = anchored ? "unbolts [src] from" : "bolts [src] to"
+	var/uraction = anchored ? "unbolt [src] from" : "bolt [src] to"
+	user.visible_message("<span class='warning'>[user] [action] the floor.</span>", "<span class='notice'>You start to [uraction] the floor...</span>", "<span class='hear'>You hear rustling noises.</span>")
+	if(W.use_tool(src, user, 50, volume=100, extra_checks = CALLBACK(src, PROC_REF(check_anchored_state), anchored)))
+		set_anchored(!anchored)
+		to_chat(user, "<span class='notice'>You [anchored ? "bolt" : "unbolt"] [src] from the floor.</span>")
+		return TRUE
+	else
+		return TRUE
```
Comment (MarkSuckerberg) [inline code/modules/mining/lavaland/ruins/gym.dm]:
```suggestion
	return TRUE
```

File: code/modules/mining/lavaland/ruins/gym.dm [hunk 9]
```diff
@@ -33,6 +80,49 @@
...
+			if(anchored)
+				return TRUE
+			to_chat(user, "<span class='notice'>You screw apart [src].</span>")
+			deconstruct(TRUE)
+		return TRUE
+
+/obj/structure/weightmachine/proc/check_anchored_state(check_anchored)
+	if(anchored != check_anchored)
+		return FALSE
+	return TRUE
```
Comment (MarkSuckerberg) [inline code/modules/mining/lavaland/ruins/gym.dm]:
```suggestion
	return anchored == check_anchored
```

**Review by MarkSuckerberg — DISMISSED** [review]
just a few tiny things

**Comment by Gristlebee** [comment]
Suggested changes have been added, and it looks like everything works

File: code/modules/mining/lavaland/ruins/gym.dm [hunk 10]
```diff
@@ -7,6 +7,47 @@
...
+			deconstruct(TRUE)
+		return TRUE
+
+/obj/structure/punching_bag/proc/check_anchored_state(check_anchored)
+	return anchored == check_anchored
+
+/obj/structure/punching_bag/examine(mob/user)
+	. = ..()
+	if(anchored)
+		. += "<span class='notice'>[src] is <b>bolted</b> to the floor.</span>"
```
Comment (thgvr) [inline code/modules/mining/lavaland/ruins/gym.dm]: Can you please convert all of these to span macros?
Example:
```suggestion
		. += span_notice("[src] is <b>bolted</b> to the floor.")
```

**Review by thgvr — CHANGES_REQUESTED** [review]

**Comment by Gristlebee** [comment]
Spans have all been converted to macros as requested

**Review by thgvr — APPROVED** [review]

**Pull request merged**